src/pj_obs_api.c

```diff
@@ -510,25 +510,198 @@ void proj_context_destroy (PJ_CONTEXT *ctx) {
+    tmpstr = getenv("HOME");
+    if (tmpstr != NULL) {
+        sprintf(info.searchpath, "%s", tmpstr);
+    }
+
+    tmpstr = getenv("PROJ_LIB");
+    if (info.searchpath != NULL) {
+        /* $HOME already in path */
+        if (tmpstr != NULL)
+            sprintf(info.searchpath, "%s;%s", info.searchpath, tmpstr);
+    } else {
+        /* path is empty */
+        if (tmpstr != NULL)
+            sprintf(info.searchpath, "%s", getenv("HOME"));
+    }
+
+    for (i=0; i<n; i++) {
+        if (strlen(info.searchpath)+strlen(paths[i]) >= 512)
+            continue;
+
+        if (info.searchpath != NULL)
+            sprintf(info.searchpath, "%s;%s", info.searchpath, paths[i]);
+        else
+            sprintf(info.searchpath, "%s", paths[i]);
+    }
+
+    return info;
 }
 
-/* ...and get rid of it safely */
-void *proj_release (void *buffer) {
-    return pj_dealloc (buffer);
+/*****************************************************************************/
+PJ_PROJ_INFO proj_pj_info(const PJ *P) {
+/******************************************************************************
+    Basic info about a particular instance of a projection object.
+
+    Returns PJ_PROJ_INFO struct.
+
+******************************************************************************/
+    PJ_PROJ_INFO info;
+    char *def;
+
+    /* projection id */
+    if (pj_param(P->ctx, P->params, "tproj").i)
+        strncpy(info.id, pj_param(P->ctx, P->params, "sproj").s, sizeof(info.id));
+
+    /* projection description */
+    strncpy(info.description, P->descr, sizeof(info.description));
+
+    /* projection definition */
+    def = pj_get_def((PJ *)P, 0); /* pj_get_def takes a non-const PJ pointer */
+    strncpy(info.definition, &def[1], sizeof(info.definition)); /* def includes a leading space */
+    pj_dealloc(def);
+
+    /* this does not take into account that a pipeline potentially does not */
+    /* have an inverse.                                                     */
+    info.has_inverse = (P->inv != 0 || P->inv3d != 0 || P->invobs != 0);
+
+    return info;
+}
+
+
+/*****************************************************************************/
+PJ_GRID_INFO proj_grid_info(const char *gridname) {
+/******************************************************************************
+    Information about a named datum grid.
+
+    Returns PJ_GRID_INFO struct.
+
+******************************************************************************/
+    PJ_GRID_INFO info;
+
+    /*PJ_CONTEXT *ctx = proj_context_create(); */
+    PJ_CONTEXT *ctx = pj_get_default_ctx();
+    PJ_GRIDINFO *gridinfo = pj_gridinfo_init(ctx, gridname);
+    memset(&info, 0, sizeof(PJ_GRID_INFO));
+
+    /* in case the grid wasn't found */
+    if (gridinfo->filename == NULL) {
+        pj_gridinfo_free(ctx, gridinfo);
+        strcpy(info.format, "missing");
+        return info;
+    }
+
+    /* name of grid */
+    strncpy(info.gridname, gridname, sizeof(info.gridname));
```

Comment:
Caution (here and in all other places where strncpy() is used). This will not zero terminate the buffer if the source argument length is >= sizeof(ret_value). I'd suggest using a proj_strlcpy() that would call strncpy() and force zero termination.